### PR TITLE
[4x] Add Multi Sorting 

### PR DESCRIPTION
**Pull request by racerguy**
Adds Multiple Column Sorting

set public $multiSort to true

Unable to test a clean pull. Please run a test if you don't mind

### CODE REVIEW

**Comment by luanfreitasdev** [comment]
Hi, thanks for submitting your PR to powergrid!.

I will be analyzing some points because there is still much to discuss.

Do you have an example showing how it works or could you submit a test?

**Review by luanfreitasdev — CHANGES_REQUESTED** [review]
There are some things in your PR that are lagging behind main and have been removed in your PR.

L:152,187,301

**Comment by racerguy** [comment]
Below is a video had to apply a fix for collections to work.

In the video I have an output at the top of the table which is not in PR below is the code for it.  

`@php
    print_r($sortArray)
@endphp`




https://user-images.githubusercontent.com/64607188/183519769-108c2668-90da-4863-8c7b-baacd305e162.mp4

**Comment by racerguy** [comment]
> There are some things in your PR that are lagging behind main and have been removed in your PR.
> 
> L:152,187,301

I believe this has been fixed in my latest commit

File: src/PowerGridComponent.php [hunk 1]
```diff
@@ -239,9 +239,21 @@ public function fillData(): mixed
...
+            foreach ($this->sortArray as $sortField => $direction) {
+                $sortField = Support\Str::of($sortField)->contains('.') || $this->ignoreTablePrefix ? $sortField : $this->currentTable . '.' . $sortField;
+                if ($this->withSortStringNumber) {
+                    $results = self::applyWithSortStringNumber($results, $sortField, $direction);
+                }
+                $results = $results->orderBy($sortField, $direction);
+            }
+        } else {
+            /* Original */
+            $sortField = Support\Str::of($this->sortField)->contains('.') || $this->ignoreTablePrefix ? $this->sortField : $this->currentTable . '.' . $this->sortField;
```
Comment (luanfreitasdev) [inline src/PowerGridComponent.php:253]: This line is duplicated with the 224

File: src/PowerGridComponent.php [hunk 2]
```diff
@@ -275,7 +287,11 @@ private function applyWithSortStringNumber(Eloquent\Builder $results, string $so
         $sortFieldType = SqlSupport::getSortFieldType($sortField);
 
         if (SqlSupport::isValidSortFieldType($sortFieldType)) {
-            $results->orderByRaw(SqlSupport::sortStringAsNumber($sortField) . ' ' . $this->sortDirection);
+            if ($multiSortDirection) {
+                $results->orderByRaw(SqlSupport::sortStringAsNumber($sortField) . ' ' . $multiSortDirection);
+            } else {
+                $results->orderByRaw(SqlSupport::sortStringAsNumber($sortField) . ' ' . $this->sortDirection);
+            }
```
Comment (luanfreitasdev) [inline src/PowerGridComponent.php:294]: Would that work?
```php
$results->orderByRaw(SqlSupport::sortStringAsNumber($sortField) . ' ' . $multiSortDirection ?:  $this->sortDirection);
```

File: src/Traits/WithSorting.php [hunk 3]
```diff
@@ -12,11 +12,19 @@ trait WithSorting
...
     {
-        $this->sortDirection = $this->sortField === $field ? $this->reverseSort() : $direction;
+        if ($this->multiSort) {
+            $this->sortByArray($field);
+        } else {
+            $this->sortDirection = $this->sortField === $field ? $this->reverseSort() : $direction;
 
-        $this->sortField = $field;
+            $this->sortField = $field;
+        }
```
Comment (luanfreitasdev) [inline src/Traits/WithSorting.php:27]:
```php
if ($this->multiSort) {
      $this->sortByArray($field);
     
      return;
}

$this->sortDirection = $this->sortField === $field ? $this->reverseSort() : $direction;

$this->sortField = $field;

File: src/Traits/WithSorting.php [hunk 4]
```diff
@@ -26,10 +34,49 @@ public function reverseSort(): string
...
+
         if (is_a($query, Collection::class)) {
             return $query->sortBy($this->sortField, SORT_REGULAR, !(($this->sortDirection === 'asc')));
         }
 
         return $query->orderBy($this->sortField, $this->sortDirection);
     }
+
+    /* change */
```
Comment (luanfreitasdev) [inline src/Traits/WithSorting.php:48]: remove comment

File: src/Traits/WithSorting.php [hunk 5]
```diff
@@ -26,10 +34,49 @@ public function reverseSort(): string
...
+                array_push($formattedSortingArray, [$sortField , $sortDirection]);
+            }
+            $query =  $query->sortBy($formattedSortingArray);
+        } else {
+            foreach ($this->sortArray as $sortField => $sortDirection) {
+                $query = $query->orderBy($sortField, $sortDirection);
+            }
+        }
+
+        return $query;
```
Comment (luanfreitasdev) [inline src/Traits/WithSorting.php:80]:
```php
if (is_a($query, Collection::class)) {
     $formattedSortingArray = [];
     foreach ($this->sortArray as $sortField => $sortDirection) {
             array_push($formattedSortingArray, [$sortField , $sortDirection]);
     }
      return  $query->sortBy($formattedSortingArray);
}

foreach ($this->sortArray as $sortField => $sortDirection) {
      $query = $query->orderBy($sortField, $sortDirection);
}

return $query;
```

**Review by luanfreitasdev — CHANGES_REQUESTED** [review]
The tests will definitely be missing at this point, but we can check the code style and phpstan for now

File: src/PowerGridComponent.php (hunk 2 above)
Comment (racerguy) [reply src/PowerGridComponent.php:294]: Changing in next PR

File: src/Traits/WithSorting.php (hunk 3 above)
Comment (racerguy) [reply src/Traits/WithSorting.php:27]: Changing in next PR

**Comment by luanfreitasdev** [comment]
Test failing is another issue with doctrine/dbal version 3.4. not related

**Comment by racerguy** [comment]
@luanfreitasdev is there any more work that needs to be completed for this to go into production?

**Comment by luanfreitasdev** [comment]
I marked it for the next version 4, because of the complexity.
I'm thinking of testing it out before merging.
Documentation is also missing.

**Comment by racerguy** [comment]
> I marked it for the next version 4, because of the complexity. I'm thinking of testing it out before merging. Documentation is also missing.

Sounds great! will try to get some basic documentation for you

**Review by luanfreitasdev — APPROVED** [review]

**Pull request merged**